CI 机器docker build均使用cache，不能发现Dockerfile新引入的问题

调查了当前合入的最新paddle版本和dockerfile上次改动前的paddle版本。
 在新搭建的CI 环境， docker build 分别出现问题。 但在已有的 CI 机器上，因为之前build成功过，这些步骤均被跳过。（using cache）

1. **最新的paddle版本**
commit 0ad9212d78d26a82e5d9722aad7807b3fede47d1 (HEAD -> test_pr, upstream/develop)
Author: Lei Wang <bestwanglei@gmail.com>
Date:   Fri May 18 19:04:13 2018 -0700

docker build 命令：
docker build **--no-cache**  -t paddlepaddle/paddle-dev:6e29fd7ace5d6cb4b88d4af864f14128 --build-arg UBUNTU_MIRROR=mirror://mirrors.ubuntu.com/mirrors.txt  --build-arg  https_proxy=http://172.19.61.250:3128  --build-arg   http_proxy=http://172.19.61.250:3128 . 

错误：
  Stored in directory: /root/.cache/pip/wheels/a9/28/53/f24776b4c5bcbe91aaf1f1e247bd6fadd17191aa12fac63902
  Running setup.py bdist_wheel for scandir: started
  Running setup.py bdist_wheel for scandir: finished with status 'error'
  Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-OOmGfn/scandir/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-jAToag --python-tag cp27:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-2.7
  copying scandir.py -> build/lib.linux-x86_64-2.7
  running build_ext
  building '_scandir' extension
  creating build/temp.linux-x86_64-2.7
  x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c _scandir.c -o build/temp.linux-x86_64-2.7/_scandir.o
  In file included from /usr/include/python2.7/Python.h:33:0,
                   from _scandir.c:14:
  /usr/include/stdio.h:33:21: fatal error: stddef.h: No such file or directory
  compilation terminated.
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for scandir
  Running setup.py clean for scandir
Successfully built nodeenv simplegeneric
Failed to build scandir
Installing collected packages: nodeenv, virtualenv, identify, cached-property, cfgv, aspy.yaml, pre-commit, wcwidth, prompt-toolkit, decorator, ptyprocess, pexpect, backports.shutil-get-terminal-size, scandir, pathlib2, ipython-genutils, enum34, traitlets, simplegeneric, pickleshare, ipython
  Running setup.py install for scandir: started
    Running setup.py install for scandir: finished with status 'error'
    Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-OOmGfn/scandir/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-Bl8sQI/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    copying scandir.py -> build/lib.linux-x86_64-2.7
    running build_ext
    building '_scandir' extension
    creating build/temp.linux-x86_64-2.7
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c _scandir.c -o build/temp.linux-x86_64-2.7/_scandir.o
    In file included from /usr/include/python2.7/Python.h:33:0,
                     from _scandir.c:14:
    /usr/include/stdio.h:33:21: fatal error: stddef.h: No such file or directory
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-OOmGfn/scandir/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-Bl8sQI/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-OOmGfn/scandir/
The command '/bin/sh -c pip install pre-commit 'ipython==5.3.0' &&     pip install 'ipykernel==4.6.0' 'jupyter==1.0.0' &&     pip install opencv-python' returned a non-zero code: 1




2. **dockerfile上次改动前的版本**
commit 8e3e65ff93718efbe3fa7f01dc52132f560e8bfc (HEAD -> test_pr)
Merge: eb7d87545e f093a7b332
Author: Xin Pan <panxin.grad@gmail.com>
Date:   Fri May 18 15:57:09 2018 +0800


docker build 命令行：
docker build **--no-cache**  -t paddlepaddle/paddle-dev:6e29fd7ace5d6cb4b88d4af864f14128 --build-arg UBUNTU_MIRROR=mirror://mirrors.ubuntu.com/mirrors.txt  --build-arg  https_proxy=http://172.19.61.250:3128  --build-arg   http_proxy=http://172.19.61.250:3128 . 
错误：
  Found existing installation: six 1.10.0
    Uninstalling six-1.10.0:
      Successfully uninstalled six-1.10.0
Exception:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/local/lib/python2.7/dist-packages/pip/commands/install.py", line 342, in run
    prefix=options.prefix_path,
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_set.py", line 795, in install
    requirement.commit_uninstall()
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_install.py", line 767, in commit_uninstall
    self.uninstalled.commit()
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_uninstall.py", line 142, in commit
    rmtree(self.save_dir)
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/retrying.py", line 49, in wrapped_f
    return Retrying(*dargs, **dkw).call(f, *args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/retrying.py", line 212, in call
    raise attempt.get()
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/retrying.py", line 247, in get
    six.reraise(self.value[0], self.value[1], self.value[2])
  File "/usr/local/lib/python2.7/dist-packages/pip/_vendor/retrying.py", line 200, in call
    attempt = Attempt(fn(*args, **kwargs), attempt_number, False)
  File "/usr/local/lib/python2.7/dist-packages/pip/utils/__init__.py", line 102, in rmtree
    onerror=rmtree_errorhandler)
  File "/usr/lib/python2.7/shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/usr/lib/python2.7/shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/usr/lib/python2.7/shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/usr/lib/python2.7/shutil.py", line 247, in rmtree
    rmtree(fullname, ignore_errors, onerror)
  File "/usr/lib/python2.7/shutil.py", line 252, in rmtree
    onerror(os.remove, fullname, sys.exc_info())
  File "/usr/local/lib/python2.7/dist-packages/pip/utils/__init__.py", line 110, in rmtree_errorhandler
    if os.stat(path).st_mode & stat.S_IREAD:
OSError: [Errno 2] No such file or directory: '/tmp/pip-SvHZmu-uninstall/usr/lib/python2.7/dist-packages/six.pyc'
You are using pip version 9.0.3, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
The command '/bin/sh -c pip install --upgrade pip==9.0.3 &&     pip install -U wheel &&     pip install -U docopt PyYAML sphinx==1.5.6 &&     pip install sphinx-rtd-theme==0.1.9 recommonmark' returned a non-zero code: 2

建议CI build 配置中加入规则:  
如果patch中 Dockerfile 文件有变更， 则需要在docker build时使用--no-cache

--- a/Dockerfile
+++ b/Dockerfile
@@ -29,7 +29,7 @@ RUN apt-get update && \
     wget unzip unrar tar xz-utils bzip2 gzip coreutils ntp \
     curl sed grep graphviz libjpeg-dev zlib1g-dev  \
     python-matplotlib gcc-4.8 g++-4.8 \
-    automake locales clang-format swig doxygen cmake  \
+    automake locales clang-format swig cmake  \
     liblapack-dev liblapacke-dev \
     clang-3.8 llvm-3.8 libclang-3.8-dev \
     net-tools libtool ccache && \

